File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    static class Node implements Comparable<Node>{
        int to;
        int cost;
        
        Node(int to, int cost){
            this.to = to;
            this.cost = cost;
        }
        
        @Override
        public int compareTo(Node other){
            return this.cost - other.cost;
        }
    }
    public int solution(int N, int[][] road, int K) {
        List<List<Node>> graph = new ArrayList<>();
        
        for(int i=0; i <= N; i++){
            graph.add(new ArrayList<>());
        }
        
        for(int[] r : road){
            int a = r[0];
            int b = r[1];
            int cost = r[2];
            
            graph.get(a).add(new Node(b, cost));
            graph.get(b).add(new Node(a, cost));
        }
        
        int[] dist = new int[N+1];
        Arrays.fill(dist, 999999);
        
        PriorityQueue<Node> pq = new PriorityQueue<>();
        
        pq.add(new Node(1,0));
        dist[1] = 0;
        
        while(!pq.isEmpty()){
            Node current = pq.poll();
            
            if(dist[current.to] < current.cost){
                continue;
            }
            
            for(Node next : graph.get(current.to)){
                if(dist[next.to] > dist[current.to] + next.cost){
                    dist[next.to] = dist[current.to] + next.cost;
                    pq.add(new Node(next.to, dist[next.to]));
                }                
            }
        }
        
        int answer = 0;
        for(int i=1; i<=N; i++){
            if(dist[i]<=K){
                answer++;
            }
        }

        // [실행] 버튼을 누르면 출력 값을 볼 수 있습니다.
        System.out.println("Hello Java");

        return answer;
    }
}

File: 프로그래머스/2/12978. 배달/배달.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <queue>
using namespace std;

int solution(int N, vector<vector<int> > road, int K) {
    vector<pair<int, int>> adj[51];
    
    for(auto r : road){
        adj[r[0]].push_back({r[1], r[2]});
        adj[r[1]].push_back({r[0], r[2]});
    }
    
    vector<int> dist(N+1, INT_MAX);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    dist[1] = 0;
    pq.push({0, 1});
    
    while(!pq.empty()){
        int d = pq.top().first;
        int curr = pq.top().second;
        pq.pop();
        
        if(d > dist[curr]){
            continue;
        }
        
        for(auto edge : adj[curr]){
            int next = edge.first;
            int cost = edge.second;
            if(dist[next] > d + cost){
                dist[next] = d + cost;
                pq.push({dist[next], next});
            }
        }
    }
    
    int answer = 0;
    for(int i = 1; i < N+1; i++){
        if(dist[i] <= K){
            answer++;
        }
    }
    
    return answer;
}